File: unixfs/fs-cursor-getter.go
package unixfs

import (
	"context"
	"sync/atomic"

	unixfs_errors "github.com/aperturerobotics/hydra/unixfs/errors"
)

// FSCursorGetter implements a FSCursor with a getter function.
// The value from the getter is returned in GetProxyCursor.
// If the getter returns nil, nil, returns nil, ErrNotExist instead.
// If the getter function is nil, returns ErrNotExist.
// The context passed to the getter should not be used after the getter returns.
// CheckReleased never returns false until Release is called.
type FSCursorGetter struct {
	released atomic.Bool
	getter   func(ctx context.Context) (FSCursor, error)
}

// NewFSCursorGetter constructs a new FSCursorGetter with a getter func.
func NewFSCursorGetter(getter func(ctx context.Context) (FSCursor, error)) *FSCursorGetter {
	return &FSCursorGetter{getter: getter}
}

// CheckReleased checks if the fscursor is released without locking anything.
func (f *FSCursorGetter) CheckReleased() bool {
	return f.released.Load()
}

// GetCursorOps returns the interface implementing FSCursorGetterOps.
// Called after AddChangeCb and only if GetProxyCursor returns nil, nil.
// Return nil, nil to indicate this position is null (nothing here).
// Return nil, ErrReleased to indicate this FSCursorGetter was released.
func (f *FSCursorGetter) GetCursorOps(ctx context.Context) (FSCursorOps, error) {
	if f.CheckReleased() {
		return nil, unixfs_errors.ErrReleased
	}
	return nil, nil
}

// GetProxyCursor returns the value from the getter, if set.
func (f *FSCursorGetter) GetProxyCursor(ctx context.Context) (FSCursor, error) {
	if f.CheckReleased() {
		return nil, unixfs_errors.ErrReleased
	}
	if f.getter == nil {
		return nil, unixfs_errors.ErrNotExist
	}
	value, err := f.getter(ctx)
	if value == nil && err == nil {
		err = unixfs_errors.ErrNotExist
	}
	return value, err
}

// Release releases the filesystem cursor.
func (f *FSCursorGetter) Release() {
	f.released.Store(true)
}

// AddChangeCb is not applicable.
func (f *FSCursorGetter) AddChangeCb(cb FSCursorChangeCb) {}

// _ is a type assertion
var _ FSCursor = ((*FSCursorGetter)(nil))

File: db/unixfs/fs-cursor-getter.ts
import type { FSCursor, FSCursorChangeCb, FSCursorOps } from './fs-cursor.js'
import { ErrReleased, ErrNotExist } from './errors/errors.js'
import type { FSHandle } from './fs-handle.js'
import { FSHandleCursor } from './fs-handle-cursor.js'

// FSCursorGetter implements a FSCursor with a getter function.
// The value from the getter is returned in getProxyCursor.
// If the getter returns null, throws ErrNotExist instead.
// If the getter function is null, throws ErrNotExist.
// checkReleased never returns true until release is called.
export class FSCursorGetter implements FSCursor {
  private released = false
  private readonly getter: ((signal?: AbortSignal) => Promise<FSCursor>) | null

  constructor(getter: ((signal?: AbortSignal) => Promise<FSCursor>) | null) {
    this.getter = getter
  }

  // checkReleased checks if the fscursor is released.
  checkReleased(): boolean {
    return this.released
  }

  // getCursorOps always returns null (this cursor always proxies).
  async getCursorOps(_signal?: AbortSignal): Promise<FSCursorOps | null> {
    if (this.released) {
      throw ErrReleased
    }
    return null
  }

  // getProxyCursor returns the value from the getter, if set.
  async getProxyCursor(signal?: AbortSignal): Promise<FSCursor | null> {
    if (this.released) {
      throw ErrReleased
    }
    if (!this.getter) {
      throw ErrNotExist
    }
    const value = await this.getter(signal)
    if (!value) {
      throw ErrNotExist
    }
    return value
  }

  // release releases the filesystem cursor.
  release(): void {
    this.released = true
  }

  // addChangeCb is not applicable.
  addChangeCb(_cb: FSCursorChangeCb): void {}

  // Symbol.dispose enables using-based cleanup.
  [Symbol.dispose](): void {
    this.release()
  }
}

// newFSCursorGetterWithHandle returns a new FSCursorGetter backed by a FSHandle.
// Constructs a FSCursor from the FSHandle when the cursor is accessed.
export function newFSCursorGetterWithHandle(handle: FSHandle): FSCursorGetter {
  return new FSCursorGetter(async (_signal?: AbortSignal) => {
    if (handle.checkReleased()) {
      throw new Error('fs cursor getter handle: cursor or inode released')
    }
    // The "false" here indicates to not release the handle.
    return new FSHandleCursor(handle, false, null)
  })
}
